wnfs/src/private/share.rs

```diff
@@ -0,0 +1,508 @@
+    use std::rc::Rc;
+
+    /// Encrypts and shares a payload with multiple recipients using their
+    /// exchange keys and stores the shares in the sharer's private forest.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn share<R: RsaKeyPair>(
+        share_payload: &SharePayload,
+        share_count: usize,
+        sharer_root_did: &str,
+        mut sharer_forest: Rc<PrivateForest>,
+        sharer_store: &mut impl BlockStore,
+        recipient_exchange_root: PublicLink,
+        recipient_store: &impl BlockStore,
+    ) -> Result<Rc<PrivateForest>> {
+        let mut exchange_keys = fetch_exchange_keys(recipient_exchange_root, recipient_store).await;
+        let encoded_payload = &dagcbor::encode(share_payload)?;
+
+        while let Some(result) = exchange_keys.next().await {
+            let exchange_key_modulus = result?;
+            let exchange_key = R::from_public_key_modulus(&exchange_key_modulus)?;
+            let encrypted_payload = exchange_key.encrypt(encoded_payload)?;
+            let share_label =
+                create_share_label(share_count, sharer_root_did, &exchange_key_modulus);
+
+            let payload_cid = sharer_store
+                .put_block(encrypted_payload, IpldCodec::Raw)
+                .await?;
+
+            sharer_forest = sharer_forest
+                .put_encrypted(share_label, payload_cid, sharer_store)
+                .await?;
+        }
+
+        Ok(sharer_forest)
+    }
+
+    /// Fetches the exchange keys of recipients using their exchange root, resolve the root_dir,
+    /// search for the exchange key, and read the exchange key's cid in the recipient's store and
+    /// yield the exchange key's value.
+    pub async fn fetch_exchange_keys(
+        recipient_exchange_root: PublicLink,
+        recipient_store: &impl BlockStore,
+    ) -> impl Stream<Item = Result<PublicKeyModulus>> + '_ {
+        Box::pin(try_stream! {
+            let root_dir = recipient_exchange_root
+                .resolve_value(recipient_store)
+                .await?
+                .as_dir()?;
+
+            let PublicOpResult { result: devices, mut root_dir  } = root_dir.ls(&[], recipient_store).await?;
+            for _ in devices {
+                let PublicOpResult { result: entries, root_dir: root } = root_dir.ls(&[], recipient_store).await?;
+                root_dir = root;
+                for (name, _) in entries {
+                    if name == EXCHANGE_KEY_NAME {
+                        root_dir
+                            .lookup_node(&name, recipient_store)
+                            .await?
+                            .ok_or(FsError::NotFound)?
+                            .as_file()?;
+
+                        let PublicOpResult { result: cid, root_dir: root } = root_dir.read(&[name], recipient_store).await?;
+                        root_dir = root;
+                        yield recipient_store.get_block(&cid).await?.to_vec();
+                        break
```

Comment:
Is this `break` necessary?

Reply:
Yes. It breaks the inner loop cos the yield won't. There is however a bug in this impl and already fixed that